### PR TITLE
[Feature/Test] Implement Model Validations and Tests

### PR DESCRIPTION
**Pull request by stino-x**
This pull request introduces comprehensive model validations and associated tests for the User, Post, Comment, and Like models. The primary focus is on ensuring data integrity, model associations, and the correct functioning of callbacks.

Changes Made
User Model:

Implemented validations for the presence of the name attribute.
Added tests for the recent_posts custom method.
Post Model:

Included validations for title presence, title length, and numericality of counters.
Implemented a callback to update the user's posts_counter after creating a post.
Added tests for validations, callbacks, and custom methods like recent_comments.
Comment Model:

Introduced a validation for the presence of the text attribute.
Implemented tests for the comment's association with a post, comments counter update, and text presence validation.
Like Model:

Added validations for the presence of user and post associations.
Implemented tests for the update_likes_counter callback.
How to Test
Run RSpec tests for each model using the following commands:
bash
Copy code
rspec spec/models/user_spec.rb
rspec spec/models/post_spec.rb
rspec spec/models/comment_spec.rb
rspec spec/models/like_spec.rb
Additional Notes
Ensure that all tests pass successfully.
If any issues are identified, please review and address them accordingly.
This pull request aims to enhance the robustness of our application by implementing thorough model validations and tests. Your feedback and reviews are highly appreciated.

### CODE REVIEW

File: db/migrate/20231108093052_create_posts.rb [hunk 1]
```diff
@@ -7,7 +7,7 @@ def change
       t.integer :likes_counter, default: 0
 
       # Add the author_id column to reference the users table
-      t.references :author, null: false, foreign_key: { to_table: :users }
```
Comment (zieeco) [inline db/migrate/20231108093052_create_posts.rb]: - Hey, kindly leave this as the author which references the user's table to maintain the ERD diagram as you've changed it currently and this doesn't follow the ERD diagram based on the project requirement.

File: db/migrate/20231111175219_change_author_id_to_user_id_in_posts.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class ChangeAuthorIdToUserIdInPosts < ActiveRecord::Migration[7.1]
+  def change
+    rename_column :posts, :author_id, :user_id
+  end
+end
```
Comment (zieeco) [inline db/migrate/20231111175219_change_author_id_to_user_id_in_posts.rb]: - After making the changes I suggested, please, kindly remove this migration file as it won't be needed anymore to meet the project requirement 
- Also, this migration file prevents your migration from running successfully, please, kindly delete it.

File: app/models/user.rb [hunk 3]
```diff
@@ -1,7 +1,10 @@
 class User < ApplicationRecord
-  has_many :posts, foreign_key: 'author_id'
+  has_many :posts, foreign_key: 'user_id'
```
Comment (zieeco) [inline app/models/user.rb]: - The same issue applies here, let's maintain the database structure as from the ERD diagram as this should be author_id. please, let's be consistent following the project requirement.

File: app/models/user.rb [hunk 4]
```diff
@@ -1,7 +1,10 @@
...
   has_many :comments
   has_many :likes
+  validates :name, presence: true
```
Comment (zieeco) [inline app/models/user.rb]: - Also kindly add a length check so users don't enter names that are too long which may cause some inconsistencies in your database. eg ` length: { maximum: 100 }`

File: db/migrate/20231108093052_create_posts.rb [hunk 5]
```diff
@@ -7,7 +7,7 @@ def change
...
+      t.references :user, null: false, foreign_key: true
```
Comment (zieeco) [inline db/migrate/20231108093052_create_posts.rb]: - You should also remove this as it doesn't follow the erd structure for the database.

File: app/models/post.rb [hunk 6]
```diff
@@ -1,11 +1,15 @@
 class Post < ApplicationRecord
-  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
```
Comment (zieeco) [inline app/models/post.rb]: - Hey, kindly leave this as the author which references the user's table to maintain the ERD diagram as you've changed it currently and this doesn't follow the ERD diagram based on the project requirement.

File: app/models/post.rb [hunk 7]
```diff
@@ -1,11 +1,15 @@
...
+  belongs_to :user, foreign_key: 'user_id'
```
Comment (zieeco) [inline app/models/post.rb]: - You should remove this also to keep your previous code which is consistent following the ERD structure for the database.

File: app/models/post.rb [hunk 8]
```diff
@@ -1,11 +1,15 @@
...
   has_many :likes
   after_create :update_user_posts_counter
+  validates :title, presence: true, length: { maximum: 250 }
+  validates :comments_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+  validates :likes_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
 
   def update_user_posts_counter
-    author.update(posts_counter: author.posts.count)
+    user.update(posts_counter: user.posts.count)
+    # user.reload
```
Comment (zieeco) [inline app/models/post.rb]: - Let's keep our previous code as it follows the ERD structure for the database.

File: db/migrate/20231111175219_change_author_id_to_user_id_in_posts.rb (hunk 2 above)
Comment (zieeco) [inline db/migrate/20231111175219_change_author_id_to_user_id_in_posts.rb]: - Please, you should remove this migration file, as you've basically changed your database structure which doesn't align with the ERD diagram from the previous milestone as this shouldn't be changed. Let's kindly update all the changes we've made.

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      user.valid?
+      expect(user.errors[:name]).to include("can't be blank")
+    end
+
+    # Uncomment the following block if you want to test bio validation
+    # it 'allows nil for bio' do
+    #   user = User.new(bio: nil)
+    #   user.valid?
+    #   expect(user.errors[:bio]).not_to include("can't be blank")
+    # end
```
Comment (zieeco) [inline spec/models/user_spec.rb]: - Let's kindly remove this commented piece of code to keep our codebase clean and maintainable.

**Review by zieeco — CHANGES_REQUESTED** [review]
Hi @stino-x,





Great job so far! 👏 
You've done well, however, there are some issues that you still need to work on to go to the next project, but you are almost there!


## To highlight 🎯

- Great effort on the model validation ✔️
- Descriptive PR. ✔️


## Required Changes ♻️
_Check the comments under the review._




## Optional suggestions

- N/A

**_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._**


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: db/migrate/20231108093052_create_posts.rb (hunk 5 above)
Comment (stino-x) [reply db/migrate/20231108093052_create_posts.rb]: please explain how

Comment (zieeco) [reply db/migrate/20231108093052_create_posts.rb]: @stino-x , [ERD DIAGRAM](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png). You have changed your code and now it doesn't follow the project structure. see the post model, it has the authorId which references the users' table. You have changed it to be userId which does follow that for the post model. Kindly update your code to meet with this erd diagram database structure for the post model.

**Review by anagudelogu — APPROVED** [review]
# STATUS: APPROVED! 🍾 🎆 🥇 

Hi @stino-x, 
I'm Andres.

You did a great job working on this milestone, your project is complete! 🚀 
There is nothing else to say other than... it's time to merge it 🤖  
Congratulations! 🎉 



## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Descriptive PR title, summary and commit messages. 💬 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by ShadyShawkat** [comment]
## Hi @stino-x, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 
- Your readme looks professional ✔️ 

Image

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**